controller/cache/cache.go

```diff
@@ -389,38 +400,125 @@ func (c *liveStateCache) Run(ctx context.Context) error {
 
 	go c.watchSettings(ctx)
 
-	kube.RetryUntilSucceed(func() error {
-		clusterEventCallback := func(event *db.ClusterEvent) {
-			c.lock.Lock()
-			cluster, ok := c.clusters[event.Cluster.Server]
-			if ok {
-				defer c.lock.Unlock()
-				if event.Type == watch.Deleted {
-					cluster.Invalidate(nil)
-					delete(c.clusters, event.Cluster.Server)
-				} else if event.Type == watch.Modified {
-					cluster.Invalidate(clustercache.SetConfig(event.Cluster.RESTConfig()))
+	localCls, err := c.db.GetCluster(ctx, common.KubernetesInternalAPIServerAddr)
```

Comment:
[ArgoDB](https://github.com/argoproj/argo-cd/blob/master/util/db/db.go) encapsulates interaction with secrets and config maps. Controller should not be aware of how exactly clusters are stored. Please move this logic into the  `WatchClusters` method. I think the updated method signature should look like this:

```golang
WatchClusters(ctx context.Context, callback func(oldCluster *appv1.Cluster, newCluster *appv1.Cluster)) error
```